File: iothermostat/mqttclient.py
#!/usr/bin/env python
# -*- coding: UTF-8 -*-

""" mqttclient.py: Provides the MqttClient class, which interfaces with the Paho
    MQTT client for the IOThermostat package """

from time import sleep
import paho.mqtt.client as mqtt
import topics
import mqttconf

__author__ = "Jan Bonne Aans"
__copyright__ = "Copyright 2018, Jan Bonne Aans"
__credits__ = []
__license__ = "GPLv3"
__version__ = "1"
__maintainer__ = "Jan Bonne Aans"
__email__ = "jbaans-at-gmail.com"
__status__ = "Development"

class MqttClient:

    data_in = {} # topics as keys and messages as values
    client = None
    subscribelist = None
    
    def __init__(self,subscribelist=None):

        print("IOThermostat: Starting MQTT client..")

        self.subscribelist = subscribelist
        client_id = 'iothermostat-backend'
        self.client = mqtt.Client(client_id=client_id)
        self.client.on_connect = self.on_connect
        self.client.on_message = self.on_message
        self.client.on_disconnect = self.on_disconnect
        self.client.will_set(topics.IOTHERMSTATUS, 'Offline', qos=0, retain=True)

        self.client.username_pw_set(mqttconf.username, password=mqttconf.password)
        self.connect()

        print("IOThermostat: MQTT client active.")


    def connect(self):
        count = 1
        connectionOk = False

        while True:
            try:
                print("IOThermostat: MQTT client connecting..")
                self.client.connect(mqttconf.server, 
                                    mqttconf.port, 
                                    mqttconf.keepalive)
                self.client.loop_start()
                connectionOk = True
                break
            except ConnectionError:
                print('IOThermostat: MQTT client connection failed {} times, retrying in 5 seconds..'.format(count))
                count += 1
                sleep(5)
                
        if not connectionOk:
            raise ConnectionError(-1, 'Error: Can\'t connect to MQTT broker.')

    def on_disconnect(self,client, userdata,rc):

        rccodes = ['0: Connection successful',
                 '1: Connection refused - incorrect protocol version',
                 '2: Connection refused - invalid client identifier',
                 '3: Connection refused - server unavailable',
                 '4: Connection refused - bad username or password',
                 '5: Connection refused - not authorised'] 
        print( "IOThermostat: MQTT client disconnected (code {}).".format(rccodes[rc]) )

    # The callback for when the client receives a CONNACK response from the server.
    def on_connect(self, client, userdata, flags, rc):

        print("IOThermostat: MQTT client connected successfully.")
        self.publish({topics.IOTHERMSTATUS:'Active'})

        print("IOThermostat: MQTT client subscribing to: ")
        for topic in self.subscribelist:
            client.subscribe(topic)
            print('IOThermostat: {}'.format(topic))


    # The callback for when a PUBLISH message is received from the server.
    def on_message(self, client, userdata, msg):
    
        # paho mqtt gives 'bXXX' or bytes. convert this to UTF-8:
        msg.payload = msg.payload.decode("utf-8")
        #print( 'MQTT client received: {}: {}'.format(msg.topic,str(msg.payload)) )
        self.data_in[msg.topic] = str(msg.payload)

    def publish(self, data):
    
        # data is a dict with keys: topics and values: ['system/sensor': 'data']
        # usage: publish({'my/topic':'my message'})
        for topic in data.keys():
            self.client.publish(topic, data[topic], mqttconf.qos, mqttconf.retain)
            #print( 'MQTT: sent: {}: {}'.format(topic,data[topic]) )

    def getData(self):
    
        # clear data after read
        result = self.data_in
        self.data_in = {}
        # return dictionary with received topics and messages
        return result

    def close(self):
    
        # closes connection
        self.client.publish(topics.IOTHERMSTATUS, 'Offline', mqttconf.qos, mqttconf.retain)
        self.client.loop_stop()
        self.client.disconnect()
        print('IOThermostat: Stopped MQTT client.')

File: iothermostat-http/scripts/mqttclient.js
/*
* Author: Jan Bonne Aans
* Copyright: Jan Bonne Aans
* License: GPLv3
* Version: 1 Development
* Contact: jbaans > gmail.com
*
* PAHO MQTT Client interface for IOThermostat
*
* - Connect to the configured MQTT server
* - Subscribe to topics
* - Publish to topics
* - Receive data
*
**/


var user;
var password;
var onconnectfun;
var onmessagefun;
var host;
var port;
var useSSL;
var mqtt;
var clientId = 'js-paho-mqttws-client-' + (Math.floor((Math.random() * 1000000) + 1));
var connected_flag = 0;
var reconnectTimeout = 15000;

// set to true to enable logging to console
var DEBUG = true;   


function floatToStr(num) {

    return num.toString().indexOf('.') === -1 ? num.toFixed(1) : num.toString();
}


function log(message){

    if (DEBUG){
        console.log(message);
    }
}


function onConnectionLost(){

    log("MQTT connection lost, reconnecting..");
    connected_flag=0;
    setTimeout(MQTTConnect, reconnectTimeout);
}


function onFailure(message) {

    log('MQTT connection failed, reconnecting..');
    setTimeout(MQTTConnect, reconnectTimeout);
}


function onMessageArrived(message){

    var topic = message.destinationName;
    var message = message.payloadString;
    log('MQTT Received: '+topic+': '+message);

    onmessagefun(topic, message);
    
}


function onConnected(recon,url){

    log(" in onConnected " +reconn);
}


function onConnect() {

    log("Connected to MQTT broker "+host +" on port "+port);
    connected_flag = 1;
    
    onconnectfun();
}


// this is a bit ugly with underscores
function MQTTSetup(_host, _port, _useSSL, _user, _password, _onconnectfun, _onmessagefun){
    host = _host;
    port = _port;
    user = _user;
    useSSL = _useSSL;
    password = _password;
    onconnectfun = _onconnectfun;
    onmessagefun = _onmessagefun;
}


function MQTTConnect(){

    log("Connecting to " + host + " on port " + port);    
    mqtt = new Paho.MQTT.Client(host,port,clientId);
    
    var options = {
        useSSL: useSSL,
        timeout: 3,
        onSuccess: onConnect,
        onFailure: onFailure,
    };

    if (typeof user !== 'undefined'){
            options.userName = user
    }

    if (typeof password !== 'undefined'){
            options.password = password
    }

    mqtt.onConnectionLost = onConnectionLost;
    mqtt.onMessageArrived = onMessageArrived;
    mqtt.onConnected = onConnected;
    mqtt.onConnect = onConnect;

    mqtt.connect(options);
}


function MQTTDisconnect(){
    mqtt.disconnect();
}


function MQTTSubscribe(topic){

    if (connected_flag==0){
        log('Not connected so can\'t subscribe');
        return false;
    }
    log("Subscribing to: "+topic);
    mqtt.subscribe(topic);
    log("Subscribed to: "+topic);
}


function MQTTPublish(topic, message){

    if (connected_flag==0){
        log('Not Connected so can\'t send');
        return false;
    }
    var qos=0;
    var retained=true;
    mqtt.send(topic,message,qos,retained);
    log("Sent: "+topic+":"+message);    
}

	
